### PR TITLE
boards: arm: Add the Wio Terminal

### PR DESCRIPTION
**Pull request by Willmish**
Wio Terminal is ATSAMD51P19 based board with an LCD Screen and
Wireless connectivity (BLE5.0 and 2.4GHz & 5GHz Wi-Fi).

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

### CODE REVIEW

**Comment by Willmish** [comment]
I fixed the documentation CI issues and refactored code to follow the required standard, however there are 3 tests run by twister that are still failing, one of which: `tests/drivers/uart/uart_async_api` I also tested for `adafruit_itsybitsy_m4_express` (same SoC) and it also fails for it, as well as some other, random boards (`atsamd21_xpro`, `adafruit_feather_m0_basic_proto`). Is it possible this test is broken? @henrikbrixandersen

**Comment by henrikbrixandersen** [comment]
> Is it possible this test is broken? @henrikbrixandersen

I'll see if I can find my `adafruit_feather_m0_basic_proto` board tonight and give it a spin. I'll get back to you.

**Comment by Willmish** [comment]
I fixed the documentation CI issues and refactored code to follow the required standard, however there are 3 tests run by twister 

> > Is it possible this test is broken? @henrikbrixandersen
> 
> I'll see if I can find my `adafruit_feather_m0_basic_proto` board tonight and give it a spin. I'll get back to you.

thanks, cause the test didnt even build.

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	rxpo = <1>;
+	txpo = <0>;
+	pinctrl-0 = <&pb27c_sercom2_pad1 &pb26c_sercom2_pad0>;
+};
+
+
+
+&dmac {
+	status = "okay";
+};
```
Comment (henrikbrixandersen) [inline boards/arm/wio_terminal/wio_terminal.dts:71]: Try leaving out this snippet from the DTS. The ASYNC UART API for the SAM0 is conditional on having an enabled DMA controller (see https://github.com/zephyrproject-rtos/zephyr/blob/aa825d77b08a4b4c1ca660dc7449517701c00227/drivers/serial/Kconfig.sam0#L15). If I enabled this on the `adafruit_feather_m0_basic_proto`, the tests fail for me as well.

Comment (Willmish) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: It still fails even without enabling 'dmac'. How do you build it for `adafruit_feather_m0_basic_proto`? I run `west build -p always -b adafruit_feather_m0_basic_proto zephyr/tests/drivers/uart/uart_async_api` and it fails the build.

Comment (henrikbrixandersen) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: Try `./scripts/twister -p adafruit_feather_m0_basic_proto`. With the above command, you force that test to be run, even though it should be filtered out for this board in `testcase.yaml`.

Comment (Willmish) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: Ok, so this test should be filtered out for` adafruit_feather_m0_basic_proto`? In this case should I also filter it out for the wio terminal? I saw that there are some board specifc overlays, like this one for `atsamd21_xpro` https://github.com/zephyrproject-rtos/zephyr/blob/aa825d77b08a4b4c1ca660dc7449517701c00227/tests/drivers/uart/uart_async_api/boards/atsamd21_xpro.overlay#L4 where the DMA channels are redefined for async operations and tx-rx are set to the same pin.

Comment (Willmish) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: Cause indeed these tests got skipped when running them with twister. (for `adafruit_feather_m0_basic_proto`)

Comment (Willmish) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: @henrikbrixandersen I also ran `./scripts/twister -p adafruit_itsybitsy_m4_express -T tests/drivers/uart/` to check if it also skips the tests for that board, but they don't get skipped and it just fails the `drivers.uart` tests,

**Comment by jadonk** [comment]
@henrikbrixandersen Do you have any additional feedback?

What does it take to get this upstream? @cfriedt any advice?

File: boards/arm/wio_terminal/pinmux.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+static int board_pinmux_init(const struct device *dev)
+{
+	__unused const struct device *muxa = DEVICE_DT_GET(DT_NODELABEL(pinmux_a));
+	__unused const struct device *muxb = DEVICE_DT_GET(DT_NODELABEL(pinmux_b));
+
+	__ASSERT_NO_MSG(device_is_ready(muxa));
+	__ASSERT_NO_MSG(device_is_ready(muxb));
+
+	ARG_UNUSED(dev);
+	ARG_UNUSED(muxa);
```
Comment (jadonk) [inline boards/arm/wio_terminal/pinmux.c:21]: I don't think muxa/muxb need to be marked ARG_UNUSED, since they aren't arguments.

**Comment by jadonk** [comment]
@Pillar1989 can you review this?

**Comment by cfriedt** [comment]
@Willmish - looks like some minor compliance issues need to be fixed. See [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow) and [Coding Style](https://docs.zephyrproject.org/latest/contribute/#coding-style).

Aside from that, for failing tests, maybe @nandojve has an idea. I'm not super familiar with this SoC and all of its quirks.

```
Run if [[ ! -s "compliance.xml" ]]; then
Error: See https://docs.zephyrproject.org/latest/contribute/#coding-style for more details.

Please remove blank lines at end of 'boards/arm/wio_terminal/wio_terminal_defconfig'
Error: See https://docs.zephyrproject.org/latest/contribute/#commit-guidelines for more details

Commit c61c1e105c:
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix documentation issue."
1: UC6 Body has no content, should at least have 1 line.

Commit 9c5e2fee5e:
1: UC6 Body has no content, should at least have 1 line.
Error: Process completed with exit code 1.
```

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		zephyr,flash = &flash0;
+		zephyr,code-partition = &code_partition;
+	};
+
+	/* These aliases are provided for compatibility with samples */
+	aliases {
+		led0 = &led0;
+        sw0 = &user_button_0;
+        sw1 = &user_button_1;
+        sw2 = &user_button_2;
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal.dts]: nit, indentation.

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        user_button_1: button_1 {
+        label = "User Button 1";
+        gpios = <&portc 27 1>;
+        };
+        user_button_2: button_2 {
+        label = "User Button 2";
+        gpios = <&portc 28 1>;
+        };
+    };
+	};
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal.dts:51]: nit, indentation.

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+&sercom2 {
+	status = "okay";
+	compatible = "atmel,sam0-uart";
+	current-speed = <115200>;
+	rxpo = <1>; 
+	txpo = <0>;
+    pinctrl-0 = <&pb27c_sercom2_pad1 &pb26c_sercom2_pad0>;
+};
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal.dts:65]: nit, indentation.

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal.dts:68]: nit, 1 line space only.

File: boards/arm/wio_terminal/wio_terminal.dts (hunk 1 above)
Comment (nandojve) [reply boards/arm/wio_terminal/wio_terminal.dts:71]: if SAM0 DMAC is enabled you need configure all DMAC channels.
There are examples at https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/uart/uart_async_api/boards

File: boards/arm/wio_terminal/wio_terminal.dts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+			reg = <0x4000 DT_SIZE_K(512-16)>;
+			read-only;
+		};
+	};
+};
+
+zephyr_udc0: &usb0 {
+    compatible = "atmel,sam0-usb";
+	status = "okay";
+};
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal.dts:96]: nit, indentation. Please, make sure you fix all indentation issues.

File: boards/arm/wio_terminal/wio_terminal_defconfig [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+# Copyright (c) 2021 Szymon Duchniewicz
+# 
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Soc/board specific
+CONFIG_SOC_SERIES_SAMD51=y
+CONFIG_SOC_PART_NUMBER_SAMD51P19A=y
+CONFIG_BOARD_WIO_TERMINAL=y
+CONFIG_SOC_ATMEL_SAMD5X_OSCULP32K_AS_MAIN=y
```
Comment (nandojve) [inline boards/arm/wio_terminal/wio_terminal_defconfig:11]: try keep CONFIG_SOC_foo together.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @Willmish,

I would recommend you squash all commits. You are adding a board and don't need fix it at same time. I recommend you run doc gen locally to fix compliance check, see [Documentation Generation](https://docs.zephyrproject.org/latest/guides/docs/index.html). You can use `make html-fast`. Make sure you follow Zephyr code style and form.

**Comment by Pillar1989** [comment]
Thanks, @jadonk and @Willmish, How about changing console output to `zephyr_udc0` as default?  @Willmish

**Comment by nandojve** [comment]
Hi @Willmish ,

Please, look at comments and let's move this after #43731.

**Comment by Willmish** [comment]
Hi @nandojve,

Had a busy time and this has slightly drifted away. Will review the comments and suggestions later this week and try to get this finished up.

**Comment by henrikbrixandersen** [comment]
@Willmish It would be great to get this merged. Any plans for picking it up again?

**Comment by Willmish** [comment]
@henrikbrixandersen hi, sorry for not being very active on this recently, but will be quite busy until end of May/beginning of June. I still want to continue working on it, but it may be a while until I have the time to dive into it again (planning to do so beginning of June).

If it is more critical to get it done now, maybe someone can pick up where I left off, otherwise I will be happy to continue working on it when my schedule clears up a bit :).

**Comment by kartben** [comment]
OK so for the record I have refreshed @Willmish's PR here https://github.com/kartben/zephyr/tree/wio_terminal_pr.
The index.rst still needs to be cleaned up but the porting itself is in a pretty good shape, and I've tested it successfuly with a number of examples.

I have added configuration for all the SERCOMs used in the Wio Terminal, which effectively allows to pre-configure things like the LCD display, accelerometer, ... in the device tree. 

I have also created a board-specific sample application that displays accelerometer data on an LVGL chart.

xxx

FWIW I am having issues with getting the SD Card to work, and I believe it's due to the sam0-spi driver not supporting `SPI_HOLD_ON_CS`, as per the discussion here: https://github.com/zephyrproject-rtos/zephyr/issues/27841.

I will open a new issue and a new PR to get the ball rolling again.

**Pull request closed without merging**